File: cpp/284.cpp
#include <vector>
using namespace std;

class Iterator {
  struct Data;
  Data *data;

public:
  Iterator(const vector<int> &nums);
  Iterator(const Iterator &iter);
  virtual ~Iterator();
  int next();
  bool hasNext() const;
};

class PeekingIterator : public Iterator {
  int peekval;
  bool hasnext;
public:
  PeekingIterator(const vector<int> &nums) : Iterator(nums) {
    // Initialize any member here.
    // **DO NOT** save a copy of nums and manipulate it directly.
    // You should only use the Iterator interface methods.
    hasnext = Iterator::hasNext();
    if(hasnext) peekval = Iterator::next();
  }

  // Returns the next element in the iteration without advancing the iterator.
  int peek() {
    return peekval;
  }

  // hasNext() and next() should behave the same as in the Iterator interface.
  // Override them if needed.
  int next() {
    int val = peekval;
    hasnext = Iterator::hasNext();
    if(hasnext) peekval = Iterator::next();
    return val;
  }

  bool hasNext() const {return hasnext;}
};

File: python/284.py
class Iterator(object):
    def __init__(self, nums):
        """
        Initializes an iterator object to the beginning of a list.
        :type nums: List[int]
        """

    def hasNext(self):
        """
        Returns true if the iteration has more elements.
        :rtype: bool
        """

    def next(self):
        """
        Returns the next element in the iteration.
        :rtype: int
        """


class PeekingIterator(object):
    def __init__(self, iterator):
        """
        Initialize your data structure here.
        :type iterator: Iterator
        """
        self.it = iterator
        self.hasnext = self.it.hasNext()
        self.peekval = self.it.next() if self.hasnext else None

    def peek(self):
        """
        Returns the next element in the iteration without advancing the iterator.
        :rtype: int
        """
        return self.peekval

    def next(self):
        """
        :rtype: int
        """
        val = self.peekval
        self.hasnext = self.it.hasNext()
        self.peekval = self.it.next() if self.hasnext else None
        return val

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.hasnext
